src/reducers/fileTree.tsx

```diff
@@ -0,0 +1,179 @@
+import { Reducer } from 'redux';
+import { ActionType, createAction, getType } from 'typesafe-actions';
+
+import { Version } from './versions';
+import { getLocalizedString } from '../utils';
+
+type FileNode = {
+  id: string;
+  name: string;
+};
+
+export type DirectoryNode = {
+  id: string;
+  name: string;
+  children: TreeNode[];
+};
+
+export type TreeNode = FileNode | DirectoryNode;
+
+const recursiveSortInPlace = (node: DirectoryNode): void => {
+  node.children.sort((a, b) => {
+    if ('children' in a && 'children' in b) {
+      return a.name.localeCompare(b.name);
+    }
+
+    if ('children' in a === false && 'children' in b === false) {
+      return a.name.localeCompare(b.name);
+    }
+
+    if ('children' in b === false) {
+      return -1;
+    }
+
+    return 1;
+  });
+
+  node.children.forEach((child) => {
+    if ('children' in child) {
+      recursiveSortInPlace(child);
+    }
+  });
+};
+
+const getVersionName = (version: Version) => {
+  return getLocalizedString(version.addon.name);
+};
+
+export const getRootPath = (version: Version) => {
+  return `root-${getVersionName(version)}`;
+};
+
+export const buildFileTree = (version: Version): DirectoryNode => {
+  const { entries } = version;
+  const root: DirectoryNode = {
+    id: getRootPath(version),
+    name: getVersionName(version),
+    children: [],
+  };
+
+  // We need to know how depth the tree is because we'll build the Treebeard
+  // tree depth by depth.
+  const maxDepth = entries.reduce((max, entry) => {
+    if (entry.depth > max) {
+      return entry.depth;
+    }
+
+    return max;
+  }, 0);
+
+  let currentDepth = 0;
+  while (currentDepth <= maxDepth) {
+    // We find the entries for the current depth.
+    const currentEntries = entries.filter(
+      // eslint-disable-next-line no-loop-func
+      (entry) => entry.depth === currentDepth,
+    );
+
+    // This is where we create new "nodes" for each entry.
+    // eslint-disable-next-line no-loop-func
+    currentEntries.forEach((entry) => {
+      let currentNode = root;
+
+      if (currentDepth > 0) {
+        // We need to find the current node (directory) to add the current
+        // entry in its children. We do this by splitting the `path` attribute
+        // and visit each node until we reach the desired node.
+        //
+        // This only applies when the current depth is not 0 (a.k.a. the root
+        // directory) because we already know `root`.
+        const parts = entry.path.split('/');
+        // Remove the filename
+        parts.pop();
+
+        for (let i = 0; i < parts.length; i++) {
+          const foundNode = currentNode.children.find(
+            (child: TreeNode) => child.name === parts[i],
+          ) as DirectoryNode;
+
+          if (foundNode) {
+            currentNode = foundNode;
+          }
+
+          // TODO: this should not happen but what if we don't find a node?
```

Comment:
I know this was in the original patch but I think it would be nice if this threw an error or logged a warning. I'm imagining that hitting this line would result in a broken tree so it would be helpful to at least get some kind of signal of what went wrong. You could probably inject a stub `root` to test it -- whatever works.